modules/apigee_edge_teams/src/Plugin/Block/TeamContextSwitcherBlock.php

```diff
@@ -0,0 +1,145 @@
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a block for switching team context.
+ *
+ * @Block(
+ *   id = "apigee_edge_team_switcher",
+ *   admin_label = @Translation("Select a team"),
+ *   category = @Translation("Apigee")
+ * )
+ */
+class TeamContextSwitcherBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $currentUser;
+
+  /**
+   * The team membership manager.
+   *
+   * @var \Drupal\apigee_edge_teams\TeamMembershipManagerInterface
+   */
+  protected $teamMembershipManager;
+
+  /**
+   * The team entity storage.
+   *
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected $storage;
+
+  /**
+   * The form builder.
+   *
+   * @var \Drupal\Core\Form\FormBuilderInterface
+   */
+  protected $formBuilder;
+
+  /**
+   * TeamContextSwitcher constructor.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Session\AccountInterface $current_user
+   *   The current user.
+   * @param \Drupal\apigee_edge_teams\TeamMembershipManagerInterface $team_membership_manager
+   *   The Apigee team membership manager.
+   * @param \Drupal\Core\Entity\EntityStorageInterface $storage
+   *   The team entity storage.
+   * @param \Drupal\Core\Form\FormBuilderInterface $form_builder
+   *   The form builder.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, AccountInterface $current_user, TeamMembershipManagerInterface $team_membership_manager, EntityStorageInterface $storage, FormBuilderInterface $form_builder) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->currentUser = $current_user;
+    $this->teamMembershipManager = $team_membership_manager;
+    $this->storage = $storage;
+    $this->formBuilder = $form_builder;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('current_user'),
+      $container->get('apigee_edge_teams.team_membership_manager'),
+      $container->get('entity_type.manager')->getStorage('team'),
+      $container->get('form_builder')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function blockAccess(AccountInterface $account) {
+    return AccessResult::allowedIf($account->isAuthenticated());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    // If the user can administer team, show all teams.
+    if ($this->currentUser->hasPermission('administer team')) {
```

Comment:
No access check should happen here. The form builder should check the available teams for a user and display only those as an option. (See my comment above).

Reply:
hmm so when a user has permission to "administer team", the user can still view all teams on the `/teams` overview page. Wouldn't it be inconsistent to be able to view all teams but not have them in the switcher?

Reply:
That could end up being a long list. Should there be a limit on the number of teams in the list?

Reply:
@arshad See my comments above regarding access checks, what is sure, that you should not check permission in the block level. All users should see this block but with different options which are handled in the form level.

@Jaesin  I do not think that a user would be a member of more than 10 groups (teams) in general which still could be displayed properly within a select list. If a customer has any problem with the default select list, s/he can install Chosen.js (to make the list searchable) or replace the form element in a form alter hook.